h/services/job_queue.py

```diff
@@ -0,0 +1,170 @@
+        self.add_sync_annotation_jobs([annotation_id], tag, scheduled_at)
+
+    def add_sync_annotation_jobs(self, annotation_ids, tag, scheduled_at=None):
+        """Add a list of annotations to the queue to be synced to Elasticsearch."""
+        self._db.add_all(
+            Job(
+                tag=tag,
+                scheduled_at=scheduled_at,
+                kwargs={"annotation_id": annotation_id},
+            )
+            for annotation_id in annotation_ids
+        )
+
+    def sync_annotations(self):
+        """
+        Synchronize annotations from Postgres to Elasticsearch.
+
+        This method is meant to be run periodically. It's called by a Celery
+        task that's scheduled periodically by h-periodic.
+
+        Each time it runs the method consumes a fixed number of sync annotation
+        jobs from the queue and copies those annotations from Postgres to
+        Elasticsearch (if they aren't in Elasticsearch already).
+
+        """
+        jobs = self._get_jobs_from_queue()
+
+        if not jobs:
+            return
+
+        annotation_ids = [job.kwargs["annotation_id"] for job in jobs]
+        annotations_from_db = self._get_annotations_from_db(annotation_ids)
+        annotations_from_es = self._get_annotations_from_es(annotation_ids)
+
+        deleted = []
+        to_index = []
+        succeeded = []
+
+        for job in jobs:
+            annotation_id = job.kwargs["annotation_id"]
+            annotation_from_db = annotations_from_db.get(annotation_id)
+            annotation_from_es = annotations_from_es.get(annotation_id)
+
+            if not annotation_from_db:
+                deleted.append(job)
+            elif annotation_from_db.deleted:
+                deleted.append(job)
+            elif not annotation_from_es:
+                to_index.append(job)
+            elif annotation_from_es["updated"] != annotation_from_db.updated:
+                to_index.append(job)
+            else:
+                succeeded.append(job)
+
+        if deleted:
+            logger.info(
+                f"Deleting {len(deleted)} sync annotation jobs because their annotations have been deleted from the DB"
+            )
+        if succeeded:
+            logger.info(
+                f"Deleting {len(succeeded)} successfully synced annotations from job queue"
+            )
+
+        for job in deleted + succeeded:
+            self._db.delete(job)
+
+        if to_index:
+            logger.info(f"Syncing {len(to_index)} annotations into Elasticsearch")
+            self._batch_indexer.index([job.kwargs["annotation_id"] for job in to_index])
+
+    def _get_jobs_from_queue(self):
+        return (
+            self._db.query(Job)
+            .filter(
+                Job.scheduled_at < datetime.datetime.utcnow(),
+            )
+            .order_by(Job.enqueued_at)
+            .limit(self._limit)
+            .with_for_update(skip_locked=True)
```

Comment:
For `FOR UPDATE` and `SKIP_LOCKED` see:

* https://www.pgcon.org/2016/schedule/attachments/414_queues-pgcon-2016.pdf
* https://www.depesz.com/2014/10/10/waiting-for-9-5-implement-skip-locked-for-row-level-locks/
* https://www.pgcasts.com/episodes/the-skip-locked-feature-in-postgres-9-5
* https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/

Basically this uses Postgres's `SELECT FOR UPDATE` to lock all the rows that we're selecting for the duration of our transaction. Then it uses Postgres's `SKIP LOCKED` to skip over any rows locked by another transaction. This means that if two instances of the service get run at the same time, they won't attempt to process the same jobs.